src/main/java/org/kizombadev/markdownparser/SyntaxAnalyzer.java

```diff
@@ -0,0 +1,161 @@
+package org.kizombadev.markdownparser;
+
+import com.google.common.collect.ImmutableList;
+import org.kizombadev.markdownparser.entities.*;
+
+import java.util.Stack;
+
+public class SyntaxAnalyzer {
+
+    private ImmutableList<Token> tokens;
+    private int tokenIndex = 0;
+    private Stack<State> stack = new Stack<>();
+
+    public static SyntaxAnalyzer create() {
+        return new SyntaxAnalyzer();
+    }
+
+    public Syntax parse(ImmutableList<Token> tokens) {
+        this.tokens = tokens;
+
+        RootSyntax root = new RootSyntax();
+
+        while (currentToken() != null) {
+            if (currentToken().equals(Token.NumberSign)) {
+                handleBigHeadlineLine(root);
+            } else if (currentToken().equals(Token.DoubleNumberSign)) {
+                handleSmallHeadlineLine(root);
+            } else if (currentToken().equals(Token.GreaterThanSign)) {
+                handleQuotation(root);
+            } else if (currentToken().equals(Token.DoubleStar)) {
+                handleBold(root);
+            } else if (currentToken().equals(Token.Star)) {
+                handleItalic(root);
+            }
+
+            if (currentToken() != null && currentToken().equals(Token.NewLine)) {
+                stepTokenForward();
+            }
+        }
+
+        return root;
+    }
+
+    private void handleQuotation(RootSyntax currentRoot) {
+        Syntax quotation = new Quotation();
+        stepTokenForward();
+
+        handleLineContainer(quotation);
+        currentRoot.addChild(quotation);
+    }
+
+    private void handleBigHeadlineLine(Syntax currentRoot) {
+        Syntax bigHeadline = new BigHeadline();
+        stepTokenForward();
+
+        handleLineContainer(bigHeadline);
+        currentRoot.addChild(bigHeadline);
+    }
+
+    private void handleSmallHeadlineLine(Syntax currentRoot) {
+        Syntax smallHeadline = new SmallHeadline();
+        stepTokenForward();
+
+
+        handleLineContainer(smallHeadline);
+        currentRoot.addChild(smallHeadline);
+    }
+
+    private void handleLineContainer(Syntax currentRoot) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS3776)